src/sage/schemes/weighted_projective/weighted_projective_space.py

```diff
@@ -0,0 +1,418 @@
+
+
+def WeightedProjectiveSpace(weights, R=None, names=None):
+    r"""
+    Return a weighted projective space with the given ``weights`` over the ring ``R``.
+
+    EXAMPLES::
+
+        sage: WP = WeightedProjectiveSpace([1, 3, 1]); WP
+        Weighted Projective Space of dimension 2 with weights (1, 3, 1) over Integer Ring
+    """
+    if (
+        isinstance(weights, (MPolynomialRing_base, PolynomialRing_generic))
+        and R is None
+    ):
+        if names is not None:
+            # Check for the case that the user provided a variable name
+            # That does not match what we wanted to use from R
+            names = normalize_names(weights.ngens(), names)
+            if weights.variable_names() != names:
+                # The provided name doesn't match the name of R's variables
+                raise NameError(
+                    "variable names passed to ProjectiveSpace conflict with names in ring"
+                )
+        A = WeightedProjectiveSpace(
+            weights.ngens() - 1, weights.base_ring(), names=weights.variable_names()
+        )
+        A._coordinate_ring = weights
+        return A
+
+    if isinstance(R, (int, Integer, list, tuple)):
+        weights, R = R, weights
+    elif R is None:
+        R = ZZ
+
+    # WeightedProjectiveSpace(5) -> just return unweighted version
+    if isinstance(weights, (int, Integer)):
+        return ProjectiveSpace(weights, R=R, names=names)
+    elif isinstance(weights, (list, tuple)):
+        # Make it hashable
+        weights = tuple(map(Integer, weights))
+        if any(w <= 0 for w in weights):
+            raise TypeError(
+                f"weights(={weights}) should only consist of positive integers"
+            )
+    else:
+        raise TypeError(f"weights={weights} must be an integer, a list or a tuple")
+
+    if names is None:
+        names = "x"
+
+    # TODO: Specialise implementation to projective spaces over non-rings.
+    if R in _CommRings:
+        return WeightedProjectiveSpace_ring(weights, R=R, names=names)
+
+    raise TypeError(f"R (={R}) must be a commutative ring")
+
+
+class WeightedProjectiveSpace_ring(UniqueRepresentation, AmbientSpace):
+    @staticmethod
+    def __classcall__(cls, weights: tuple[Integer], R=ZZ, names=None):
+        # __classcall_ is the "preprocessing" step for UniqueRepresentation
+        # see docs of CachedRepresentation
+        # weights should be a tuple, also because it should be hashable
+        if not isinstance(weights, tuple):
+            raise TypeError(
+                f"weights(={weights}) is not a tuple. Please use the"
+                "`WeightedProjectiveSpace` constructor"
+            )
+
+        normalized_names = normalize_names(len(weights), names)
+        return super().__classcall__(cls, weights, R, normalized_names)
+
+    def __init__(self, weights: tuple[Integer], R=ZZ, names=None):
+        """
+        Initialization function.
```

Comment:
Add a more detailed docstring.

Reply:
Not worth it, nobody is seeing the docstring of `__init__` instead of that of the class.